### PR TITLE
Simplification rule for overflowing additions.

### LINKED ISSUE #1061: Investigate `tests/specs/mcd/flipper-addu48u48-fail-rough-spec.k`
This test is failing in 180s or so on Haskell backend, it's a good one to look at.

### PR DESCRIPTION
**Pull request by RaoulSchaffranek**
Fixes #1061

### CODE REVIEW

File: tests/specs/mcd/functional-spec.k [hunk 1]
```diff
@@ -27,6 +27,11 @@ module FUNCTIONAL-SPEC
        andBool Art_i <=Int maxSInt256
        andBool #rangeSInt(256, Art_i *Int ABI_rate)
 
+    claim <k> runLemma(maxUInt48 &Int ABI_x <=Int maxUInt48 &Int ABI_x +Int ABI_y) => doneLemma(false) ... </k>
+      requires #rangeUInt(48, ABI_x)
+        andBool #rangeUInt(48, ABI_y)
+        andBool notBool #rangeUInt(48, ABI_x +Int ABI_y)
```
Comment (ehildenb) [inline tests/specs/mcd/functional-spec.k:33]:
```suggestion
      requires #rangeUInt(48, ABI_x)
       andBool #rangeUInt(48, ABI_y)
       andBool notBool #rangeUInt(48, ABI_x +Int ABI_y)
```

File: tests/specs/mcd/verification.k [hunk 2]
```diff
@@ -85,7 +85,10 @@ module LEMMAS-MCD-HASKELL [symbolic, kore]
     rule #range(M:Memory [ K <- V ]                        , START, WIDTH) => #range(M [ K := (V : .ByteArray) ], START, WIDTH)  requires K +Int 1                    ==Int START +Int WIDTH andBool START <=Int K andBool 0 <Int WIDTH andBool 0 <=Int START [concrete(K    , V     ), simplification]
     rule #range(M:Memory [ K <- V ] [ K' := VS':ByteArray ], START, WIDTH) => #range(M [ K := (V : VS')        ], START, WIDTH)  requires K' +Int #sizeByteArray(VS') ==Int START +Int WIDTH andBool START <=Int K andBool 0 <Int WIDTH andBool 0 <=Int START
                                                                                                                                   andBool K +Int 1                    ==Int K'                                                                                [concrete(K, K', V, VS'), simplification]
-
+    rule maxUInt48 &Int X <=Int maxUInt48 &Int X +Int Y => false
+      requires #rangeUInt(48, X)
+        andBool #rangeUInt(48, Y)
+        andBool notBool #rangeUInt(48, X +Int Y) [simplification]
```
Comment (ehildenb) [inline tests/specs/mcd/verification.k:91]:
```suggestion
      requires #rangeUInt(48, X)
       andBool #rangeUInt(48, Y)
       andBool notBool #rangeUInt(48, X +Int Y) [simplification]
```

**Comment by ehildenb** [comment]
@RaoulSchaffranek says that this proof takes 1 hour to pass now, and we're spending 40 minutes in Z3. We should inspect the Z3 queries we're making for (i) redundancy (are we making the smae query multiple times), and (ii) necessity (are we sending things to Z3 we colud simplify with K rules).

**Comment by RaoulSchaffranek** [comment]
With 39a5bcc the test now passes in under 6m compared to roughly 1h before the change.

**Comment by Riib11** [comment]
@RaoulSchaffranek are you actively working on this rn? I can take a look at it to see why it isn't passing tests.

**Comment by RaoulSchaffranek** [comment]
Hi @Riib11, thanks for pointing out to me that the CI pipeline is failing. I will investigate.

File: tests/specs/mcd/verification.k [hunk 3]
```diff
@@ -78,19 +78,46 @@ module LEMMAS-MCD-HASKELL [symbolic, kore]
...
+  // ########################
+
+    rule <k> LT maxUInt48 &Int X +Int Y maxUInt48 &Int X => 1 ~> #push ... </k>
+      requires #rangeUInt(48, X)
+       andBool #rangeUInt(48, Y)
+       andBool notBool #rangeUInt(48, X +Int Y) [simplification]
+
+    rule <k> ADD X Y => X +Int Y ~> #push ... </k>
+      requires #rangeUInt(48, X)
+       andBool #rangeUInt(48, Y) [simplification]
```
Comment (ehildenb) [inline tests/specs/mcd/verification.k:105]: We can't merge rule replacements like this, they are not very safe for updates to the semantics.

Comment (ehildenb) [reply tests/specs/mcd/verification.k:105]: Instead, add enough functional `runLemma => doneLemma` style claims that each of the new lemmas you add is tested, and then we don't have to enable this test on CI until th ebackend issues that prompted these additions are resolved.

Comment (RaoulSchaffranek) [reply tests/specs/mcd/verification.k]: I used these rules to workaround a performance bottleneck in the Haskell backend. Now that we have [fixed the original performance issue](https://github.com/kframework/kore/pull/2800) directly in the backend, these rules are no longer needed.

With the workaround, the test needed roughly 6m to succeed on my machine. With the workaround removed and the proper fix applied, performance is even better at roughly 3~4m.

**Review by ehildenb — APPROVED** [review]

**Pull request merged**